File: sdks/android/src/test/kotlin/com/rynl10n/ScenarioTest.kt
package com.rynl10n

import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/** DoD ① — 시나리오 A/B/C를 public API(RynL10nClient)로 재현. */
class ScenarioTest {
    private fun snap(release: String, base: String, locales: Map<String, Map<String, TranslationValue>>) =
        Snapshot(1, release, base, "en", locales)
    private fun text(s: String) = TranslationValue.Text(s)

    @Test fun scenarioA_otaHotfix() {
        val v0 = snap("R42", "base0", mapOf("en" to mapOf("pay.button" to text("Pay")), "ja" to mapOf("pay.button" to text("支払―"))))
        val delta = Delta(1, "R42", "base0", "base1", listOf(DeltaOp("set", "pay.button", "ja", text("支払い"))))
        val store = InMemoryDeliveryStore()
        store.put(v0, "releases/R42/snapshot-base0.json")
        store.put(delta, "releases/R42/delta-base0-base1.json")

        val published = ManifestRelease("R42", ReleaseState.PUBLISHED,
            VersionMatch("semver-range", ">=3.2.0 <3.3.0"), "base0", "base1", 100,
            "releases/R42/snapshot-base0.json", "releases/R42/delta-base0-base1.json")
        val manifest = Manifest(1, "shop", "en", "T1", listOf(published))

        val client = RynL10nClient(v0, store, Matching.ClientContext(appVersion = "3.2.1"))
        var notified = 0
        client.onCatalogUpdated { notified++ }

        assertEquals("支払―", client.t("pay.button", locale = "ja"))
        assertTrue(client.refresh(manifest))
        assertEquals("支払い", client.t("pay.button", locale = "ja"))
        assertEquals(1, notified)

        val rolledBack = ManifestRelease("R42", ReleaseState.PUBLISHED,
            VersionMatch("semver-range", ">=3.2.0 <3.3.0"), "base0", "base0", 100,
            "releases/R42/snapshot-base0.json", null)
        client.refresh(Manifest(1, "shop", "en", "T2", listOf(rolledBack)))
        assertEquals("支払―", client.t("pay.button", locale = "ja"))
    }

    @Test fun scenarioB_deterministicBake() {
        val a = buildJsonObject {
            put("en", buildJsonObject { put("greet", JsonPrimitive("Hello")) })
            put("ko", buildJsonObject { put("greet", JsonPrimitive("안녕하세요")) })
        }
        val b = buildJsonObject {
            put("ko", buildJsonObject { put("greet", JsonPrimitive("안녕하세요")) })
            put("en", buildJsonObject { put("greet", JsonPrimitive("Hello")) })
        }
        val ha = ContentHash.snapshotHash("R1", "en", a)
        assertEquals(ha, ContentHash.snapshotHash("R1", "en", b))
        assertEquals(16, ContentHash.fileId(ha).length)
        // NFC: 조합형으로 들어와도 같은 해시
        val decomposed = buildJsonObject {
            put("en", buildJsonObject { put("greet", JsonPrimitive("Hello")) })
            put("ko", buildJsonObject { put("greet", JsonPrimitive(java.text.Normalizer.normalize("안녕하세요", java.text.Normalizer.Form.NFD))) })
        }
        assertEquals(ha, ContentHash.snapshotHash("R1", "en", decomposed))
    }

    @Test fun scenarioC_versionIsolation() {
        val r42 = snap("R42", "r42", mapOf("en" to mapOf("home.title" to text("Home"))))
        val r50 = snap("R50", "r50", mapOf("en" to mapOf("home.title" to text("Home"), "home.newBadge" to text("NEW"))))
        val store = InMemoryDeliveryStore()
        store.put(r42, "releases/R42/snapshot-r42.json")
        store.put(r50, "releases/R50/snapshot-r50.json")

        val releases = listOf(
            ManifestRelease("R42", ReleaseState.SUPERSEDED, VersionMatch("semver-range", ">=3.2.0 <3.3.0"),
                "r42", "r42", 100, "releases/R42/snapshot-r42.json", null),
            ManifestRelease("R50", ReleaseState.PUBLISHED, VersionMatch("semver-range", ">=3.3.0"),
                "r50", "r50", 100, "releases/R50/snapshot-r50.json", null),
        )
        val manifest = Manifest(1, "app", "en", "T", releases)

        val oldApp = RynL10nClient(r42, store, Matching.ClientContext(appVersion = "3.2.5"))
        oldApp.refresh(manifest)
        assertEquals("R42", oldApp.status().releaseId)
        assertEquals("Home", oldApp.t("home.title"))
        assertEquals("⟪home.newBadge⟫", oldApp.t("home.newBadge"))

        val newApp = RynL10nClient(r50, store, Matching.ClientContext(appVersion = "3.3.1"))
        newApp.refresh(manifest)
        assertEquals("R50", newApp.status().releaseId)
        assertEquals("NEW", newApp.t("home.newBadge"))
    }

    @Test fun rangeConflictDetection() {
        val overlapping = listOf(
            Matching.ConflictInput("R42", VersionMatch("semver-range", ">=3.2.0 <3.4.0")),
            Matching.ConflictInput("R60", VersionMatch("semver-range", ">=3.3.0 <3.5.0")),
        )
        assertEquals(1, Matching.findRangeConflicts(overlapping).size)
        val adjacent = listOf(
            Matching.ConflictInput("R42", VersionMatch("semver-range", ">=3.2.0 <3.3.0")),
            Matching.ConflictInput("R50", VersionMatch("semver-range", ">=3.3.0")),
        )
        assertEquals(0, Matching.findRangeConflicts(adjacent).size)
    }
}

File: sdks/flutter/test/scenario_test.dart
// DoD ① — 시나리오 A/B/C를 public API(RynL10nClient)로 재현.
import 'package:test/test.dart';
import 'package:rynl10n/rynl10n.dart';

Snapshot snap(String release, String base, Map<String, Map<String, String>> locales) {
  final tv = locales.map((loc, keys) =>
      MapEntry(loc, keys.map((k, v) => MapEntry(k, TextValue(v) as TranslationValue))));
  return Snapshot(1, release, base, 'en', tv);
}

void main() {
  test('시나리오 A — OTA 오타 긴급 수정 + 롤백', () {
    final v0 = snap('R42', 'base0', {'en': {'pay.button': 'Pay'}, 'ja': {'pay.button': '支払―'}});
    final delta = Delta('R42', 'base0', 'base1', [DeltaOp('set', 'pay.button', 'ja', TextValue('支払い'))]);
    final store = InMemoryDeliveryStore();
    store.putSnapshot('releases/R42/snapshot-base0.json', v0);
    store.putDelta('releases/R42/delta-base0-base1.json', delta);

    final published = ManifestRelease('R42', 'published', VersionMatch('semver-range', '>=3.2.0 <3.3.0'),
        'base0', 'base1', 100, 'releases/R42/snapshot-base0.json', 'releases/R42/delta-base0-base1.json');
    final manifest = Manifest('shop', 'en', [published]);

    final client = RynL10nClient(bundle: v0, store: store, context: ClientContext(appVersion: '3.2.1'));
    var notified = 0;
    client.onCatalogUpdated((_) => notified++);

    expect(client.t('pay.button', locale: 'ja'), '支払―');
    expect(client.refresh(manifest), true);
    expect(client.t('pay.button', locale: 'ja'), '支払い');
    expect(notified, 1);

    final rolledBack = ManifestRelease('R42', 'published', VersionMatch('semver-range', '>=3.2.0 <3.3.0'),
        'base0', 'base0', 100, 'releases/R42/snapshot-base0.json', null);
    client.refresh(Manifest('shop', 'en', [rolledBack]));
    expect(client.t('pay.button', locale: 'ja'), '支払―');
  });

  test('시나리오 B — 결정적 bake(같은 카탈로그 → 같은 base, NFC 무관)', () {
    final a = {'en': {'greet': 'Hello'}, 'ko': {'greet': '안녕하세요'}};
    final b = {'ko': {'greet': '안녕하세요'}, 'en': {'greet': 'Hello'}};
    final ha = snapshotHash('R1', 'en', a);
    expect(snapshotHash('R1', 'en', b), ha); // 키 순서 무관
    expect(fileId(ha).length, 16);
  });

  test('시나리오 C — 앱 버전별 격리(신규 키가 구버전에 안 샘)', () {
    final r42 = snap('R42', 'r42', {'en': {'home.title': 'Home'}});
    final r50 = snap('R50', 'r50', {'en': {'home.title': 'Home', 'home.newBadge': 'NEW'}});
    final store = InMemoryDeliveryStore();
    store.putSnapshot('releases/R42/snapshot-r42.json', r42);
    store.putSnapshot('releases/R50/snapshot-r50.json', r50);

    final releases = [
      ManifestRelease('R42', 'superseded', VersionMatch('semver-range', '>=3.2.0 <3.3.0'), 'r42', 'r42', 100, 'releases/R42/snapshot-r42.json', null),
      ManifestRelease('R50', 'published', VersionMatch('semver-range', '>=3.3.0'), 'r50', 'r50', 100, 'releases/R50/snapshot-r50.json', null),
    ];
    final manifest = Manifest('app', 'en', releases);

    final oldApp = RynL10nClient(bundle: r42, store: store, context: ClientContext(appVersion: '3.2.5'));
    oldApp.refresh(manifest);
    expect(oldApp.releaseId, 'R42');
    expect(oldApp.t('home.newBadge'), '⟪home.newBadge⟫');

    final newApp = RynL10nClient(bundle: r50, store: store, context: ClientContext(appVersion: '3.3.1'));
    newApp.refresh(manifest);
    expect(newApp.releaseId, 'R50');
    expect(newApp.t('home.newBadge'), 'NEW');
  });

  test('카나리 게이트: rollout 0 → 오버레이 미수신', () {
    final v0 = snap('R1', 'b0', {'en': {'greet': 'old'}});
    final delta = Delta('R1', 'b0', 'b1', [DeltaOp('set', 'greet', 'en', TextValue('new'))]);
    final store = InMemoryDeliveryStore();
    store.putDelta('releases/R1/delta-b0-b1.json', delta);
    final rel = ManifestRelease('R1', 'published', VersionMatch('semver-range', '>=1.0.0'), 'b0', 'b1', 0, 'releases/R1/snapshot-b0.json', 'releases/R1/delta-b0-b1.json');
    final client = RynL10nClient(bundle: v0, store: store, context: ClientContext(appVersion: '1.0.0'), installId: 'device-1');
    client.refresh(Manifest('p', 'en', [rel]));
    expect(client.t('greet'), 'old'); // rollout 0 → 미수신
  });

  test('정수 빌드 넘버 매칭', () {
    final releases = [
      ManifestRelease('B1', 'published', VersionMatch('integer-range', '>=42 <50'), 'b1', 'b1', 100, 's1', null),
      ManifestRelease('B2', 'published', VersionMatch('integer-range', '>=50'), 'b2', 'b2', 100, 's2', null),
    ];
    expect(selectRelease(releases, ClientContext(buildNumber: 45)).releaseId, 'B1');
    expect(selectRelease(releases, ClientContext(buildNumber: 60)).releaseId, 'B2');
  });
}
